### PR TITLE
Allow to "break" sensors

### PR DESCRIPTION
**Pull request by jbmouret**
When we split a skeleton, some sensors point to unavailable bodies (or the robot would have no access to the sensors on the detached skeleton).

This PR use the `_attached` boolean (which was already here) to be able detach a sensor explicitly (to avoid a segv when it was attached to something that does not exist anymore).

One idea is that it is up to the user to know what to do when a sensor is detached (e.g., replace with zero). So, the user should check that the sensor is attached when working with damage and querying sensors.

### CODE REVIEW

File: src/robot_dart/robot.cpp [hunk 1]
```diff
@@ -1618,6 +1622,11 @@ namespace robot_dart {
         return detail::dof_data<15>(_skeleton, dof_names, _dof_map);
     }
 
+    Eigen::VectorXd Robot::constraint_forces(const std::vector<std::string>& dof_names) const
```
Comment (costashatz) [inline src/robot_dart/robot.cpp:1625]: What does the `constraint_forces` has to do with the detaching of sensor (that this PR handles)? If this is needed for something else, can you make a different PR? Thanks!

File: src/robot_dart/sensor/sensor.cpp [hunk 2]
```diff
@@ -43,8 +43,18 @@ namespace robot_dart {
         void Sensor::set_pose(const Eigen::Isometry3d& tf) { _world_pose = tf; }
         const Eigen::Isometry3d& Sensor::pose() const { return _world_pose; }
 
+        void Sensor::detach() {
+            _attached_to_body = false;
+            _attached_to_joint = false;
+            _body_attached = NULL;
```
Comment (costashatz) [inline src/robot_dart/sensor/sensor.cpp:49]: Can you use `nullptr` instead of `NULL`?

File: src/robot_dart/sensor/sensor.cpp [hunk 3]
```diff
@@ -111,5 +120,13 @@ namespace robot_dart {
                 _attached_to_joint = false;
             }
         }
+        const std::string& Sensor::attached_to() const
+         { 
+            assert(_attached_to_body || _attached_to_joint);
```
Comment (costashatz) [inline src/robot_dart/sensor/sensor.cpp:125]: Please use robot_dart's assertion macros. See [here](https://github.com/resibots/robot_dart/blob/master/src/robot_dart/utils.hpp#L48-L76).

**Review by costashatz — CHANGES_REQUESTED** [review]
@jbmouret thanks a lot for this. See my comments above.

File: src/robot_dart/sensor/sensor.cpp (hunk 2 above)
Comment (jbmouret) [reply src/robot_dart/sensor/sensor.cpp:49]: sure, old habits :)

File: src/robot_dart/robot.cpp (hunk 1 above)
Comment (jbmouret) [reply src/robot_dart/robot.cpp:1625]: We need it to compute the SPD+ (for position control) with a broken robot. I agree that this is not about "broken sensor", but we need broken sensors to handle damaged robots. I don't think few lines are worth a PR. We can change the commit message if you care.

File: src/robot_dart/sensor/sensor.cpp (hunk 3 above)
Comment (costashatz) [reply src/robot_dart/sensor/sensor.cpp]: Thanks a ton!

File: src/robot_dart/robot.cpp (hunk 1 above)
Comment (costashatz) [reply src/robot_dart/robot.cpp:1625]: OK then. Then change the name of the PR (e.g. `Allow to "break" sensors + constraint forces`) or put in bold in the description something about this..

**Review by costashatz — APPROVED** [review]

**Comment by costashatz** [comment]
@jbmouret @dalinel thanks for this!
